crates/task-impls/src/consensus/view_change.rs

```diff
@@ -0,0 +1,132 @@
+use std::sync::Arc;
+
+use crate::{
+    events::HotShotEvent,
+    helpers::{broadcast_event, cancel_task},
+};
+use anyhow::{ensure, Result};
+use async_broadcast::Sender;
+use async_compatibility_layer::art::{async_sleep, async_spawn};
+use async_lock::{RwLock, RwLockUpgradableReadGuard};
+use core::time::Duration;
+use hotshot_types::{
+    consensus::Consensus,
+    constants::LOOK_AHEAD,
+    traits::{
+        election::Membership,
+        network::{ConnectedNetwork, ConsensusIntentEvent},
+        node_implementation::{ConsensusTime, NodeImplementation, NodeType},
+    },
+};
+use tracing::{debug, error};
+
+#[cfg(async_executor_impl = "async-std")]
+use async_std::task::JoinHandle;
+#[cfg(async_executor_impl = "tokio")]
+use tokio::task::JoinHandle;
+
+/// Update the view if it actually changed, takes a mutable reference to the `cur_view` and the
+/// `timeout_task` which are updated during the operation of the function.
+///
+/// # Errors
+/// Returns an [`anyhow::Error`] when the new view is not greater than the current view.
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn update_view<TYPES: NodeType, I: NodeImplementation<TYPES>>(
+    public_key: TYPES::SignatureKey,
+    new_view: TYPES::Time,
+    event_stream: &Sender<Arc<HotShotEvent<TYPES>>>,
+    quorum_membership: Arc<TYPES::Membership>,
+    quorum_network: Arc<I::QuorumNetwork>,
+    timeout: u64,
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+    cur_view: &mut TYPES::Time,
+    timeout_task: &mut Option<JoinHandle<()>>,
```

Comment:
Question: what was the reason for changing this function from being a member of the consensus task state struct?

Reply:
We need to be able to use it within the `QuorumProposalRecv` task. In order to process that specific event, we need to call `update_view` at some point, but we cannot without inheriting a ton of the baggage of the consensus task.